app/app.go

```diff
@@ -784,3 +785,26 @@ func (app *App) InitializeAppVersion(ctx sdk.Context) {
 		app.SetAppVersion(ctx, appVersion)
 	}
 }
+
+// OfferSnapshot is a wrapper around the baseapp's OfferSnapshot method. It is
+// needed to mount stores for the appropriate app version.
+func (app *App) OfferSnapshot(req abci.RequestOfferSnapshot) abci.ResponseOfferSnapshot {
+	if app.IsSealed() {
+		// If the app is sealed, keys have already been mounted so this can
+		// delegate to the baseapp's OfferSnapshot.
+		return app.BaseApp.OfferSnapshot(req)
+	}
+
+	if app.upgradeHeightV2 == 0 {
+		app.Logger().Debug("v2 upgrade height not set, assuming app version 2")
+		app.mountKeysAndInit(v2)
+	} else if req.Snapshot.Height >= uint64(app.upgradeHeightV2) {
+		app.Logger().Debug("snapshot height is greater than or equal to upgrade height, assuming app version 2")
+		app.mountKeysAndInit(v2)
+	} else {
+		app.Logger().Debug("snapshot height is less than upgrade height, assuming app version 1")
+		app.mountKeysAndInit(v1)
+	}
```

Comment:
@cmwaters is it necessary to also invoke something like:

```
			app.SetAppVersion(ctx, v2)
```

here? When I tried that, I got a panic because

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x8 pc=0x10585c2e0]

goroutine 135 [running]:
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).NewContext(_, _, {{0x0, 0x0}, {0x0, 0x0}, 0x0, {0x0, 0x0, 0x0}, ...})
	/Users/rootulp/git/rootulp/celestiaorg/cosmos-sdk/baseapp/test_helpers.go:46 +0x70
github.com/celestiaorg/celestia-app/v2/app.(*App).OfferSnapshot(0x140007dd508, {0x14001266b40?, {0x140010be540?, 0x104c7cbac?, 0x1?}})
	/Users/rootulp/git/rootulp/celestiaorg/celestia-app/app/app.go:794 +0x6c
github.com/tendermint/tendermint/abci/client.(*localClient).OfferSnapshotSync(0x14001c003c0, {0x14001266b40?, {0x140010be540?, 0x107597fc0?, 0x1?}})
	/Users/rootulp/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.40.0-tm-v0.34.29/abci/client/local_client.go:330 +0x110
github.com/tendermint/tendermint/proxy.(*appConnSnapshot).OfferSnapshotSync(0x1400035c0a0?, {0x14001266b40?, {0x140010be540?, 0x140013bfec0?, 0x20?}})
	/Users/rootulp/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.40.0-tm-v0.34.29/proxy/app_conn.go:199 +0x30
github.com/tendermint/tendermint/statesync.(*syncer).offerSnapshot(0x14001710240, 0x140005b9da0)
	/Users/rootulp/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.40.0-tm-v0.34.29/statesync/syncer.go:325 +0x220
github.com/tendermint/tendermint/statesync.(*syncer).Sync(_, _, _)
	/Users/rootulp/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.40.0-tm-v0.34.29/statesync/syncer.go:269 +0x178
github.com/tendermint/tendermint/statesync.(*syncer).SyncAny(_, _, _)
	/Users/rootulp/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.40.0-tm-v0.34.29/statesync/syncer.go:185 +0x3f0
github.com/tendermint/tendermint/statesync.(*Reactor).Sync(_, {_, _}, _)
	/Users/rootulp/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.40.0-tm-v0.34.29/statesync/reactor.go:303 +0x2f8
github.com/tendermint/tendermint/node.startStateSync.func1()
	/Users/rootulp/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.40.0-tm-v0.34.29/node/node.go:725 +0x90
created by github.com/tendermint/tendermint/node.startStateSync in goroutine 14
	/Users/rootulp/go/pkg/mod/github.com/celestiaorg/celestia-core@v1.40.0-tm-v0.34.29/node/node.go:724 +0x31c
```

Reply:
Looks like the panic is from how you initialise the context.

Reply:
But yes, we should be setting the app version here

Reply:
Turns out we don't because the state that we sync has the app version and when we call `Info` we update it from the store

Reply:
Callum tested offline and since `Info` gets invoked after state sync and app version is set there, it doesn't need to be set here.